deepspeed/runtime/zero/partition_parameters.py

```diff
@@ -3,29 +3,68 @@
 Licensed under the MIT license.
 """
 
+import math
 import os
 import time
 import types
+from typing import Callable, Iterable
 from enum import Enum
 import functools
 import itertools
+from typing import List
 
 import torch
-from torch.distributed.distributed_c10d import _get_global_rank, group
+from torch import Tensor
 import torch.distributed as dist
+from torch.distributed.distributed_c10d import _get_global_rank, group
+from torch.nn import Module
+from torch.nn import Parameter
 
 from .linear import LinearModuleForZeroStage3, LinearFunctionForZeroStage3
 from .offload_constants import *
 
-from ..utils import see_memory_usage
-from deepspeed.utils import log_dist, init_distributed, logger
+from ..utils import get_only_unique_item, see_memory_usage
+from deepspeed.runtime.zero.utils import assert_ints_same_as_other_ranks
+from deepspeed.utils import init_distributed, instrument_w_nvtx, logger
 from deepspeed.utils.debug import debug_param2name_id_shape, debug_param2name_id_shape_device, debug_module2name, debug_param2name, debug_param2name_id_shape_status, printflock, log_rank_file
+from deepspeed.utils.logging import logger
 
 from ..swap_tensor.partitioned_param_swapper import AsyncPartitionedParameterSwapper, PartitionedParamStatus
 from ..config import DeepSpeedConfig
 
 param_count = 0
-partitioned_param_data_shape = [1]
+partitioned_param_data_shape = [0]
+
+if hasattr(torch.distributed, "_all_gather_base"):
```

Comment:
Could we avoid duplication by import/reuse this logic from [here](https://github.com/jfc4050/DeepSpeed/blob/2b5f6ea2bba91ea1a5e5b9d2febedf041926d810/deepspeed/runtime/comm/coalesced_collectives.py#L16-L38)?

Reply:
sorry i dont fully understand, can you clarify?

Reply:
I mean these codes in [partition_parameters.py](https://github.com/jfc4050/DeepSpeed/blob/2b5f6ea2bba91ea1a5e5b9d2febedf041926d810/deepspeed/runtime/zero/partition_parameters.py#L38-L67) and [coalesced_collectives.py](https://github.com/jfc4050/DeepSpeed/blob/2b5f6ea2bba91ea1a5e5b9d2febedf041926d810/deepspeed/runtime/comm/coalesced_collectives.py#L16-L38) are quite similar, or perhaps I am missing the subtle difference. So I was wondering if we could avoid the code duplication.

Reply:
there's some subtle differences, i'm not sure how if i know how to reuse without complicating/obfuscating things. can you suggest a change?

Reply:
In that case, don't worry about it for now. We can always revisit later.